presto-native-execution/presto_cpp/main/TaskManager.cpp

```diff
@@ -588,56 +588,58 @@ struct ZombieTaskCounts {
   folly::F14FastSet<protocol::TaskId> taskIdsToClean;
-  { taskMap = *(taskMap_.rlock()); }
 
   ZombieTaskCounts zombieTaskCounts;
   ZombieTaskCounts zombiePrestoTaskCounts;
-  for (auto it = taskMap.begin(); it != taskMap.end(); ++it) {
-    bool eraseTask{false};
-    if (it->second->task != nullptr) {
-      if (it->second->task->state() != exec::TaskState::kRunning) {
-        if (it->second->task->timeSinceEndMs() >= FLAGS_old_task_ms) {
-          // Not running and old.
+  {
+    // We copy task map locally to avoid locking task map for a potentially long
+    // time. We also lock for 'read'.
+    TaskMap taskMap = *(taskMap_.rlock());
+
+    for (auto it = taskMap.begin(); it != taskMap.end(); ++it) {
+      bool eraseTask{false};
+      if (it->second->task != nullptr) {
+        if (it->second->task->state() != exec::TaskState::kRunning) {
+          if (it->second->task->timeSinceEndMs() >= FLAGS_old_task_ms) {
+            // Not running and old.
+            eraseTask = true;
+          }
+        }
+      } else {
+        // Use heartbeat to determine the task's age.
+        if (it->second->timeSinceLastHeartbeatMs() >= FLAGS_old_task_ms) {
           eraseTask = true;
         }
       }
-    } else {
-      // Use heartbeat to determine the task's age.
-      if (it->second->timeSinceLastHeartbeatMs() >= FLAGS_old_task_ms) {
-        eraseTask = true;
-      }
-    }
 
-    // We assume 'not erase' is the 'most common' case.
-    if (not eraseTask) {
-      continue;
-    }
+      // We assume 'not erase' is the 'most common' case.
+      if (not eraseTask) {
+        continue;
+      }
 
-    auto prestoTaskRefCount = it->second.use_count();
-    auto taskRefCount = it->second->task.use_count();
-
-    // Do not remove 'zombie' tasks (with outstanding references) from the map.
-    // We use it to track the number of tasks.
-    // Note, since we copied the task map, presto tasks should have an extra
-    // reference (2 from two maps).
-    if (prestoTaskRefCount > 2 || taskRefCount > 1) {
-      auto& task = it->second->task;
-      if (prestoTaskRefCount > 2) {
-        ++zombiePrestoTaskCounts.numTotal;
-        if (task != nullptr) {
-          zombiePrestoTaskCounts.updateCounts(task);
+      auto prestoTaskRefCount = it->second.use_count();
+      auto taskRefCount = it->second->task.use_count();
+
+      // Do not remove 'zombie' tasks (with outstanding references) from the
+      // map. We use it to track the number of tasks. Note, since we copied the
+      // task map, presto tasks should have an extra reference (2 from two
+      // maps).
+      if (prestoTaskRefCount > 2 || taskRefCount > 1) {
+        auto& task = it->second->task;
+        if (prestoTaskRefCount > 2) {
+          ++zombiePrestoTaskCounts.numTotal;
+          if (task != nullptr) {
+            zombiePrestoTaskCounts.updateCounts(task);
+          }
         }
+        if (taskRefCount > 1) {
+          ++zombieTaskCounts.numTotal;
```

Comment:
Consider to rename it to zombieVeloxTaskCounts?

And rename kCounterNumZombieTasks to kCounterNumZombieVeloxTasks and the corresponding stats name to make it clear? Thanks!